File: sdks/python/src/oap_sdk/aht.py
"""Ad Hoc Teamwork (RFC 0027 revision 2).

Helpers to build and verify Capability Announcements, drive the Three-Tier
Convention Discovery Handshake, and evaluate AHT Fallback Policies.

The Three-Tier algorithm is the load-bearing element of revision 2: it
makes the protocol unilaterally adoptable (Theorem A.3) by extending
revision 1's explicit Schelling reduction (Tier 1) with Bayesian
observational inference (Tier 2) and minimax-regret robust selection
(Tier 3) over the joint posterior. When all peers are protocol-followers,
Tier 3 collapses to the Tier 1 result.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Literal, Optional

from .signing import canonicalize, sha256_hex

AhtPolicyClass = Literal["POAM", "PLASTIC", "AATEAM", "ROTATE", "Custom"]
PeerClass = Literal["P", "O", "A"]


@dataclass
class AhtFallbackPolicy:
    policy_class: AhtPolicyClass
    assumptions: List[str]
    policy_ref: Optional[str] = None
    training_distribution_ref: Optional[str] = None


@dataclass
class Peer:
    did: str
    classification: PeerClass
    convention_space: Optional[List[Any]] = None
    observed_actions: Optional[List[Any]] = None


@dataclass
class ThreeTierParams:
    unilateral_timeout_ms: int = 1500
    regret_tolerance: float = 0.10
    max_byzantine_fraction: float = 0.0


@dataclass
class ThreeTierResult:
    committed_convention: Optional[Any]
    tier_used: Literal["tier1", "tier1+3", "tier2+3", "fallback-only", "abort"]
    posterior: Dict[str, Dict[str, float]] = field(default_factory=dict)
    worst_case_regret: Optional[float] = None
    reason: Optional[str] = None
    receipts: List[Any] = field(default_factory=list)
    action: Optional[Any] = None


def aht_canonicalize(value: Any) -> str:
    return canonicalize(value)


def capability_announcement_hash(announcement: Dict[str, Any]) -> str:
    return "sha256:" + sha256_hex(aht_canonicalize(announcement))


def _bayesian_posterior(
    observations: List[Any],
    type_space: List[str],
    action_likelihood: Callable[[Any, str], float],
) -> Dict[str, float]:
    n = len(type_space) or 1
    post = {th: 1.0 / n for th in type_space}
    for a in observations:
        z = 0.0
        for th in type_space:
            post[th] = post[th] * max(action_likelihood(a, th), 1e-12)
            z += post[th]
        if z > 0:
            for th in type_space:
                post[th] /= z
    return post


def _expected_regret_under_posterior(
    convention: Any,
    posterior: Dict[str, float],
    regret: Callable[[Any, str], float],
) -> float:
    return sum(p * regret(convention, th) for th, p in posterior.items())


def run_three_tier_handshake(
    *,
    self_agent: Dict[str, Any],
    peers: List[Peer],
    fallback_policy: Callable[[List[Any], Dict[str, Dict[str, float]]], Any],
    action_likelihood: Callable[[Any, str], float],
    type_space: List[str],
    regret: Callable[[Any, str], float],
    convention_space_for_type: Callable[[str], List[Any]],
    params: ThreeTierParams,
) -> ThreeTierResult:
    """Three-Tier Convention Discovery Handshake (RFC 0027 section 3.4).

    Theorem A.1 (Unilateral Bounded Termination) holds for any
    |N_P| >= 0, including the boundary case in which no peers publish.
    """
    n = peers + [Peer(did=self_agent["did"], classification="P", convention_space=self_agent["convention_space"])]
    n_p = [p for p in n if p.classification == "P"]
    n_o = [p for p in peers if p.classification == "O"]
    n_a = [p for p in peers if p.classification == "A"]

    t = len(n_a)
    if len(n) < 3 * t + 1:
        return ThreeTierResult(committed_convention=None, tier_used="abort", reason="byzantine-bound-violated")

    # Tier 1
    provisional: Optional[Any] = None
    if n_p:
        sets = [set(map(aht_canonicalize, p.convention_space or [])) for p in n_p]
        if sets:
            inter = sets[0]
            for s in sets[1:]:
                inter = inter & s
            if inter:
                provisional_key = sorted(inter)[0]
                import json
                provisional = json.loads(provisional_key)

    # Tier 2
    posterior: Dict[str, Dict[str, float]] = {}
    for j in n_o:
        posterior[j.did] = _bayesian_posterior(j.observed_actions or [], type_space, action_likelihood)

    # Tier 3
    feasible: set[str] = set()
    if provisional is not None:
        max_r = 0.0
        for j in n_o:
            r = _expected_regret_under_posterior(provisional, posterior[j.did], regret)
            if r > max_r:
                max_r = r
        if max_r <= params.regret_tolerance:
            feasible.add(aht_canonicalize(provisional))
            if not n_o and not n_a:
                return ThreeTierResult(committed_convention=provisional, tier_used="tier1", posterior=posterior)

    if not feasible:
        for j in n_o:
            for th in [t for t, p in posterior[j.did].items() if p > 0]:
                for c in convention_space_for_type(th):
                    feasible.add(aht_canonicalize(c))
        for c in self_agent.get("convention_space", []) or []:
            feasible.add(aht_canonicalize(c))

    committed: Optional[Any] = None
    best_worst_case = math.inf
    import json as _json
    for c_key in feasible:
        c = _json.loads(c_key)
        per_peer_regret: List[float] = []
        for j in [p for p in n if p.did != self_agent["did"]]:
            r = 0.0
            if j.classification == "P":
                published = list(map(aht_canonicalize, j.convention_space or []))
                r = 0.0 if c_key in published else 1.0
            elif j.classification == "O":
                r = _expected_regret_under_posterior(c, posterior[j.did], regret)
            elif j.classification == "A":
                r = 1.0
            per_peer_regret.append(r)
        per_peer_regret.sort(reverse=True)
        trimmed = per_peer_regret[t:]
        worst_case = trimmed[0] if trimmed else 0.0
        if worst_case < best_worst_case:
            best_worst_case = worst_case
            committed = c

    if committed is None:
        action = fallback_policy([], posterior)
        return ThreeTierResult(committed_convention=None, tier_used="fallback-only", posterior=posterior, action=action)

    tier_used: Literal["tier1+3", "tier2+3"]
    if provisional is not None and aht_canonicalize(provisional) == aht_canonicalize(committed):
        tier_used = "tier1+3"
    else:
        tier_used = "tier2+3"
    return ThreeTierResult(committed_convention=committed, tier_used=tier_used, posterior=posterior, worst_case_regret=best_worst_case)


def detect_convention_drift(
    *,
    posterior: Dict[str, float],
    recent_empirical: Dict[str, float],
    threshold_kl: float,
) -> Dict[str, Any]:
    """RFC 0027 section 3.4b. Returns dict with 'drifted' bool and 'kl_divergence' float."""
    kl = 0.0
    for th, p in recent_empirical.items():
        q = posterior.get(th, 1e-12)
        if p > 0:
            kl += p * math.log(p / q)
    return {"drifted": kl > threshold_kl, "kl_divergence": kl}

File: sdks/typescript/src/aht.ts
/**
 * Ad Hoc Teamwork (RFC 0027 revision 2).
 *
 * Helpers to build and verify Capability Announcements, drive the Three-Tier
 * Convention Discovery Handshake, and evaluate AHT Fallback Policies.
 *
 * The Three-Tier algorithm is the load-bearing element of revision 2: it
 * makes the protocol unilaterally adoptable (Theorem A.3) by extending
 * revision 1's explicit Schelling reduction (Tier 1) with Bayesian
 * observational inference (Tier 2) and minimax-regret robust selection
 * (Tier 3) over the joint posterior. When all peers are protocol-followers,
 * Tier 3 collapses to the Tier 1 result.
 */

import { canonicalize, sha256Hex } from './signing.js';

export type AhtPolicyClass = 'POAM' | 'PLASTIC' | 'AATEAM' | 'ROTATE' | 'Custom';

export type PeerClass = 'P' | 'O' | 'A';

export interface CapabilityAnnouncement {
  schema: 'oap.capability.v1';
  agent_did: string;
  context: { context_type: 'workflow' | 'session'; context_id: string; step_id?: string | null };
  capabilities: Array<{
    action: string;
    schema_ref: string;
    preconditions?: string[];
    cost_class?: 'free' | 'billable' | 'stakeable';
    expected_latency_ms?: number;
  }>;
  evidence: {
    manifest_url: string;
    reputation_profile_ref?: string;
    role_enactment?: { organization_did: string; role_id: string; role_credential?: string };
    byoa_attestation_ref?: string;
  };
  issued_at?: string;
  expires_at?: string;
  signature: string;
}

export interface AhtFallbackPolicy {
  policy_class: AhtPolicyClass;
  policy_ref?: string;
  assumptions: string[];
  training_distribution_ref?: string;
}

export interface Peer<C extends PeerClass = PeerClass> {
  did: string;
  classification: C;
  conventionSpace?: unknown[];
  observedActions?: unknown[];
}

export interface ThreeTierParams {
  unilateralTimeoutMs: number;
  regretTolerance: number;
  maxByzantineFraction: number;
}

export interface ThreeTierResult {
  committedConvention: unknown | null;
  tierUsed: 'tier1' | 'tier1+3' | 'tier2+3' | 'fallback-only' | 'abort';
  posterior: Record<string, Record<string, number>>;
  worstCaseRegret?: number;
  reason?: string;
  receipts: unknown[];
  action?: unknown;
}

export interface ThreeTierInputs {
  self: { did: string; conventionSpace: unknown[] };
  peers: Peer[];
  fallbackPolicy: (history: unknown[], posterior: Record<string, Record<string, number>>) => unknown;
  actionLikelihood: (action: unknown, theta: string) => number;
  typeSpace: string[];
  regret: (convention: unknown, theta: string) => number;
  conventionSpaceForType: (theta: string) => unknown[];
  params: ThreeTierParams;
}

export function ahtCanonicalize(value: unknown): string {
  return canonicalize(value as Record<string, unknown>);
}

export async function capabilityAnnouncementHash(a: CapabilityAnnouncement): Promise<string> {
  return 'sha256:' + sha256Hex(ahtCanonicalize(a));
}

function bayesianPosterior(
  observations: unknown[],
  typeSpace: string[],
  actionLikelihood: (a: unknown, theta: string) => number,
): Record<string, number> {
  const post: Record<string, number> = {};
  const n = typeSpace.length;
  for (const th of typeSpace) post[th] = 1 / n;
  for (const a of observations) {
    let z = 0;
    for (const th of typeSpace) {
      post[th] = post[th] * Math.max(actionLikelihood(a, th), 1e-12);
      z += post[th];
    }
    if (z > 0) for (const th of typeSpace) post[th] /= z;
  }
  return post;
}

function expectedRegretUnderPosterior(
  convention: unknown,
  posterior: Record<string, number>,
  regret: (c: unknown, theta: string) => number,
): number {
  let s = 0;
  for (const th of Object.keys(posterior)) s += posterior[th] * regret(convention, th);
  return s;
}

/**
 * Three-Tier Convention Discovery Handshake (RFC 0027 section 3.4).
 *
 * Tier 1 runs Schelling reduction over publishers; Tier 2 runs Bayesian
 * inference over silent peers; Tier 3 selects via t-byzantine-trimmed
 * minimax regret. Theorem A.1 (Unilateral Bounded Termination) holds for
 * any |N_P| >= 0, including the boundary case in which no peers publish.
 */
export function runThreeTierHandshake(opts: ThreeTierInputs): ThreeTierResult {
  const { self, peers, fallbackPolicy, actionLikelihood, typeSpace, regret, conventionSpaceForType, params } = opts;

  const N = [...peers, self];
  const NP = N.filter((p) => (p as Peer).classification === 'P' || p === self);
  const NO = peers.filter((p) => p.classification === 'O');
  const NA = peers.filter((p) => p.classification === 'A');

  const t = NA.length;
  if (N.length < 3 * t + 1) {
    return { committedConvention: null, tierUsed: 'abort', posterior: {}, reason: 'byzantine-bound-violated', receipts: [] };
  }

  // Tier 1
  let provisional: unknown = null;
  if (NP.length >= 1) {
    const sets = NP.map((p) => new Set(((p as Peer).conventionSpace || []).map(ahtCanonicalize)));
    const inter = sets.length === 0 ? [] : [...sets[0]].filter((k) => sets.every((s) => s.has(k)));
    if (inter.length > 0) {
      inter.sort();
      provisional = JSON.parse(inter[0]);
    }
  }

  // Tier 2
  const posterior: Record<string, Record<string, number>> = {};
  for (const j of NO) {
    posterior[j.did] = bayesianPosterior(j.observedActions || [], typeSpace, actionLikelihood);
  }

  // Tier 3
  const F = new Set<string>();
  if (provisional !== null) {
    let max = 0;
    for (const j of NO) {
      const r = expectedRegretUnderPosterior(provisional, posterior[j.did], regret);
      if (r > max) max = r;
    }
    if (max <= params.regretTolerance) {
      F.add(ahtCanonicalize(provisional));
      if (NO.length === 0 && NA.length === 0) {
        return { committedConvention: provisional, tierUsed: 'tier1', posterior, receipts: [] };
      }
    }
  }
  if (F.size === 0) {
    for (const j of NO) {
      const supportedTypes = Object.keys(posterior[j.did]).filter((th) => posterior[j.did][th] > 0);
      for (const th of supportedTypes) for (const c of conventionSpaceForType(th)) F.add(ahtCanonicalize(c));
    }
    for (const c of self.conventionSpace || []) F.add(ahtCanonicalize(c));
  }

  let committed: unknown = null;
  let bestWorstCase = Infinity;
  for (const cKey of F) {
    const c = JSON.parse(cKey);
    const perPeerRegret: number[] = [];
    for (const j of N.filter((p) => p !== self)) {
      let r = 0;
      const jp = j as Peer;
      if (jp.classification === 'P') {
        const published = (jp.conventionSpace || []).map(ahtCanonicalize);
        r = published.includes(cKey) ? 0 : 1;
      } else if (jp.classification === 'O') {
        r = expectedRegretUnderPosterior(c, posterior[jp.did], regret);
      } else if (jp.classification === 'A') {
        r = 1;
      }
      perPeerRegret.push(r);
    }
    perPeerRegret.sort((a, b) => b - a);
    const trimmed = perPeerRegret.slice(t);
    const worstCase = trimmed.length > 0 ? trimmed[0] : 0;
    if (worstCase < bestWorstCase) {
      bestWorstCase = worstCase;
      committed = c;
    }
  }

  if (committed === null) {
    const action = fallbackPolicy([], posterior);
    return { committedConvention: null, tierUsed: 'fallback-only', posterior, action, receipts: [] };
  }

  const tierUsed: ThreeTierResult['tierUsed'] =
    provisional !== null && ahtCanonicalize(provisional) === ahtCanonicalize(committed) ? 'tier1+3' : 'tier2+3';
  return { committedConvention: committed, tierUsed, posterior, worstCaseRegret: bestWorstCase, receipts: [] };
}

export function detectConventionDrift(opts: {
  posterior: Record<string, number>;
  recentEmpirical: Record<string, number>;
  thresholdKl: number;
}): { drifted: boolean; klDivergence: number } {
  let kl = 0;
  for (const th of Object.keys(opts.recentEmpirical)) {
    const p = opts.recentEmpirical[th];
    const q = opts.posterior[th] ?? 1e-12;
    if (p > 0) kl += p * Math.log(p / q);
  }
  return { drifted: kl > opts.thresholdKl, klDivergence: kl };
}
